File: src/main/java/ar/com/flow/bankaccount/transaction/Builder.java
package ar.com.flow.bankaccount.transaction;

import ar.com.flow.money.Money;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class Builder {
    private Collection<Precondition> preconditions = new ArrayList<>();
    private List<Step> steps = new ArrayList<>();
    private Money amount;

    public Builder precondition(Precondition precondition) {
        preconditions.add(precondition);
        return this;
    }

    public Builder step(Step step) {
        steps.add(step);
        return this;
    }

    public Builder amount(Money amount) {
        this.amount = amount;
        return this;
    }

    public Transaction build() {
        return new Transaction(new Preconditions(preconditions), steps, amount);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/transaction/Builder.kt
package ar.com.flow.bankaccount.transaction

import ar.com.flow.money.Money

class Builder {
    private val preconditions: MutableList<Precondition> = ArrayList()
    private val steps: MutableList<Step> = ArrayList()
    private lateinit var amount: Money

    fun precondition(precondition: Precondition): Builder {
        preconditions.add(precondition)
        return this
    }

    fun step(step: Step): Builder {
        steps.add(step)
        return this
    }

    fun amount(amount: Money): Builder {
        this.amount = amount
        return this
    }

    fun build(): Transaction {
        return Transaction(preconditions, steps, amount)
    }
}